pkg/configstore/store.go

```diff
@@ -0,0 +1,179 @@
+}
+
+func listFiles(dirname string) ([]string, error) {
+	files, err := ioutil.ReadDir(dirname)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := make([]string, len(files))
+	for idx, file := range files {
+		ret[idx] = file.Name()
+	}
+
+	return ret, nil
+}
+
+func parseFile(path string, value interface{}) error {
+	reader, err := os.Open(path)
+	defer reader.Close()
+	if err != nil {
+		return fmt.Errorf("Failed to open %s: %v", path, err)
+	}
+
+	return json.NewDecoder(reader).Decode(value)
+}
+
+func fileExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// NewStore creates a new config store
+// `dir` is a directory from where data will be loaded
+// `hints` is a dictionary to find which file to load for
+// a given key
+func NewStore(dir string, hints map[string]string) *Store {
+	if hints == nil {
+		hints = map[string]string{}
+	}
+	return &Store{
+		Dir:        dir,
+		Hints:      hints,
+		ListFiles:  listFiles,
+		ParseFile:  parseFile,
+		FileExists: fileExists,
+	}
+}
+
+// Decode `path` contents using `json`, lock the store mutex, merge loaded data
+// into s.Data, unlock the mutex
+func (s *Store) loadPath(path string) error {
+	value := map[string]interface{}{}
+	err := s.ParseFile(path, value)
+	if err != nil {
+		return fmt.Errorf("Failed to parse %s: %v", path, err)
+	}
+
+	s.Lock()
+	defer s.Unlock()
+	for key, val := range value {
+		s.Data[key] = val
+	}
+
+	return nil
+}
+
+// Walk `s.dir` and `s.loadPath` all the files
+func (s *Store) loadAll() error {
+	files, err := s.ListFiles(s.Dir)
+	if err != nil {
+		return fmt.Errorf("Failed to list %s: %v", s.Dir, err)
+	}
+
+	for _, file := range files {
+		filepath := path.Join(s.Dir, file)
+		err := s.loadPath(filepath)
+		if err != nil {
+			return fmt.Errorf("Failed to load %s: %v", filepath, err)
+		}
+	}
+	return nil
+}
+
+var extensions = []string{"json", "yaml"}
+
+// Look for `file`.json or `file`.yaml, if not found try loading all files and
+// print a warning about hints
+func (s *Store) load(file string) error {
+	for _, ext := range extensions {
+		path := path.Join(s.Dir, fmt.Sprintf("%s.%s", file, ext))
+		exists, err := s.FileExists(path)
+		if err != nil {
+			return fmt.Errorf("Failed to find %s: %v", path, err)
+		}
+		if exists {
+			err = s.loadPath(path)
+			if err != nil {
+				return fmt.Errorf("Failed to load %s: %v", path, err)
+			}
+			return nil
+		}
+	}
+
+	log.Printf(
+		"WARN: loading all configs, consider adding some hints in %s",
+		path.Join(s.Dir, file),
+	)
+	return s.loadAll()
+}
+
+func (s *Store) syncGet(key string) (interface{}, bool) {
+	s.Lock()
+	defer s.Unlock()
+	val, ok := s.Data[key]
+	return val, ok
+}
+
+// Get returns value for given `key`. If not found in `s.data`, call
+// `s.load` function with `file` from `s.hints` or `key` itself.
+func (s *Store) Get(key string) (interface{}, error) {
```

Comment:
so when you call this, you either call this with a key, or the filename itself? in the instance that the key has a different name to the file, do you then have to make an additional call to get the contents of the key?

it's kind of surprising to work like this to be honest - wouldn't it be more 'obvious' if you had a function to load the contents of a specific config file, or get a key from any file in a specific directory?

the previous way of doing this, even if expensive, hides the details of the filesystem layout. we're very much pushing that back to the user this way, aren't we?

Reply:
you always call for the key, never filename. if key is to be found in a different filename than `<key>.json`, this mapping has to be provided in `hints` argument when config store object is created. contents of resolved file are merged into `configstore.Data`, so no additional lookup for the same key is required by user.

it's a tradeoff i made to make interface of loading paasta configs similar to what we have in python paasta, so that migrating code is less burdensome. the fallback case (eager loading when no direct correspondence betwee key and filename is found and no hint is provided) makes it (almost?) completely compatible with how configs work in paasta.